### PR TITLE
Iam authentication

### PR DESCRIPTION
**Pull request by SpacemanPaul**
### Reason for this pull request

Cloud ops want to be able to authenticate to RDS databases using IAM authentication

### Proposed changes

1. Utility function in `utils/aws/__init__.py` to obtain a new IAM authentication token.
2. Hook in Postgresql driver to register a `do_connect` event on newly instantiated SQLAlchemy engines, to obtain a new IAM token every 10 minutes.

Written so as to easily support other dynamic password-generation frameworks.

To use: Set Environment Variable `$ODC_IAM_RDS_AUTHENTICATION` to 'yes'.

### CODE REVIEW

**Comment by SpacemanPaul** [comment]
Not sure if just randomly reading an environment variable is the right thing to do.  Should this really be bundled up with the rest of the AWS configuration or something?

**Comment by omad** [comment]
Hi @jeremyh , can you please have a look at this. I'm not so familiar with the connection handling.

File: datacube/utils/aws/__init__.py [hunk 1]
```diff
@@ -438,3 +438,15 @@ def get_aws_settings(profile: Optional[str] = None,
                  aws_secret_access_key=cc.secret_key,
                  aws_session_token=cc.token,
                  requester_pays=requester_pays), creds)
+
+
+def obtain_new_IAM_auth_token(url: str, region_name: str = "auto", profile_name: Optional[str] = None) -> str:
```
Comment (jeremyh) [inline datacube/utils/aws/__init__.py]: - `url: URL`, not str
- lowercase function name for consistency with the rest of ODC (I don't think there's strict pep8 linter enabled, but it's convention elsewhere in the repo I believe)

**Review by jeremyh — COMMENTED** [review]
This looks good. 

My only suggestion is that I think we should keep it together with our other config and environment handling rather than in a separate part of the codebase.

The current ones are in [LocalConfig](https://github.com/opendatacube/datacube-core/blob/3f9a7fe8950dad8f057d6ebda6b3c1a025e2c96e/datacube/config.py#L60-L62), and all fields there can be defined in one conf file and kept together. And the other envionment variables are handled in that class too.

(See [DATACUBE_ENVIRONMENT](https://github.com/opendatacube/datacube-core/blob/91e0f0b78fa7aec2947c3cd6a12e5716edbb4dfa/datacube/config.py#L87), and [DATACUBE_DB_URL](https://github.com/opendatacube/datacube-core/blob/91e0f0b78fa7aec2947c3cd6a12e5716edbb4dfa/datacube/config.py#L195) as examples of existing env vars -- both of which override fields that exist in the original config)

Minor things:
- Prefer `DATACUBE_` prefix instead of `ODC_` for consistency with other env variables. Maybe `DATACUBE_IAM_AUTHENTICATION`, as in theory a future index may not be RDS (?).
- Ideally make that auth token refresh time a config option too while we're here -- it feels inevitable that different environments will want to customise it (?)

File: datacube/utils/aws/__init__.py (hunk 1 above)
Comment (SpacemanPaul) [reply datacube/utils/aws/__init__.py]: Thanks that's really helpful.

**Comment by SpacemanPaul** [comment]
How's this?

File: tests/test_geometry.py [hunk 2]
```diff
@@ -1541,6 +1541,22 @@ def test_lonlat_bounds():
     with pytest.raises(ValueError):
         geometry.lonlat_bounds(geometry.box(0, 0, 1, 1, None))
 
+    multi = {
```
Comment (SpacemanPaul) [inline tests/test_geometry.py:1544]: I don't know why this change is here. I think it's from a different (already merged) PR via rebase weirdness.

File: datacube/utils/geometry/_base.py [hunk 3]
```diff
@@ -1389,35 +1389,30 @@ def lonlat_bounds(geom: Geometry,
     if geom.crs.geographic:
         return geom.boundingbox
```
Comment (SpacemanPaul) [inline datacube/utils/geometry/_base.py:1391]: These changes are not mine.  I think something went wrong with the rebase.  See also below.

**Comment by SpacemanPaul** [comment]
Hang on, the branch has got a bit tangled with merges and rebases.  I'll need a bit longer to cleanup.
